hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -387,31 +387,37 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 	}
 
 	reportLimitedSuportEnabled(hcluster)
-	reportSilecedAlerts(hcluster)
+	reportSilencedAlerts(hcluster)
 	reportProxyConfig(hcluster)
 
 	// If deleted, clean up and return early.
 	if !hcluster.DeletionTimestamp.IsZero() {
 		reportHostedClusterGuestCloudResourcesDeletionDuration(hcluster)
+		reportHostedClusterDeletionDuration(hcluster, r.Clock)
 
 		// Keep trying to delete until we know it's safe to finalize.
 		completed, err := r.delete(ctx, hcluster)
 		if err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to delete hostedcluster: %w", err)
 		}
+
 		if !completed {
 			log.Info("hostedcluster is still deleting", "name", req.NamespacedName)
 			return ctrl.Result{RequeueAfter: clusterDeletionRequeueDuration}, nil
 		}
+
 		// Now we can remove the finalizer.
 		if controllerutil.ContainsFinalizer(hcluster, finalizer) {
 			controllerutil.RemoveFinalizer(hcluster, finalizer)
 			if err := r.Update(ctx, hcluster); err != nil {
+				hcmetrics.StopReporting(hcluster)
```

Comment:
why is this being called here? if err !=nil then the finaliser is not removed

Reply:
It is because of the following weird case / race condition I experimented while testing:
1. A first call to `reconcile` removes the finalizer.
2. This changes the `hostedcluster` in DB which is not yet deleted...  
so a second call to `reconcile` is started
3. The `hcluster` variable is set and filled from the `hostedcluster` object read from the DB
4. The`hostedcluster` object is deleted from the DB (because there is no remaining finalizer)
5. Some `report` methods is called and set the metrics for the deleted `hostedcluster` again
6. We try to delete the finalizer again, but this time this will fail as `hostedcluster` no longer exist in DB.  
**At that time we have to make sure that the metrics eventually "recreated" again at step #5 are ended again.**